solidity/contracts/AttributeCheckpointFraud.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+
+import {TREE_DEPTH} from "./libs/Merkle.sol";
+import {CheckpointLib, Checkpoint} from "./libs/CheckpointLib.sol";
+import {CheckpointFraudProofs} from "./CheckpointFraudProofs.sol";
+
+enum FraudType {
+    NOT_PROVEN,
+    Whitelist,
+    Premature,
+    MessageId,
+    Root
+}
+
+/**
+ * @title AttributeCheckpointFraud
+ * @dev The AttributeCheckpointFraud contract is used to attribute fraud to a specific ECDSA checkpoint signer.
+ */
+contract AttributeCheckpointFraud is Ownable {
+    using CheckpointLib for Checkpoint;
+    using Address for address;
+
+    CheckpointFraudProofs public immutable checkpointFraudProofs =
+        new CheckpointFraudProofs();
+
+    mapping(address => bool) public merkleTreeWhitelist;
+
+    mapping(address signer => mapping(bytes32 digest => FraudType))
+        public attributions;
+
+    function whitelist(address merkleTree) external onlyOwner {
```

Comment:
do we need a unwhitelist function too?

Reply:
if we add then validators will need to monitor this contract, imo its unnecessary